### DISCUSSION TITLE
`Broadcast` is called after `NotifyCanExecuteChanged`

### DISCUSSION THREAD

**Discussion by hellow554** [CommunityToolkit/dotnet#770] [Q&A]
I have the following Property and Command

```csharp
[ObservableProperty]
[NotifyCanExecuteChangedFor(nameof(ExportButtonCommand))]
[NotifyPropertyChangedRecipients]
private string exportPath = string.Empty;

[RelayCommand(CanExecute = nameof(CanExport))]
private void ExportButton() { }

private bool CanExport() { return ExportService.CanExport(); }
```

ExportService looks like this:

```csharp
class ExportService : ObservableRecipient, IRecipient<PropertyChangedMessage<string>> {
    private string? Outputpath { get; set; }
    public bool CanExport() {
        return !string.IsNullOrEmpty(Outputpath);
    }
    public void Receive(PropertyChangedMessage<string> message) {
        if (message.PropertyName == nameof(ExportViewModel.ExportPath)) {
            Outputpath = message.NewValue;
        }
    }
}
```

---

Now, the problem is, the following generated Code:

```csharp
public string ExportPath
{
    get => exportPath;
    [global::System.Diagnostics.CodeAnalysis.MemberNotNull("exportPath")]
    set
    {
        if (!global::System.Collections.Generic.EqualityComparer<string>.Default.Equals(exportPath, value))
        {
            string __oldValue = exportPath;
            OnExportPathChanging(value);
            OnExportPathChanging(default, value);
            OnPropertyChanging(global::CommunityToolkit.Mvvm.ComponentModel.__Internals.__KnownINotifyPropertyChangingArgs.ExportPath);
            exportPath = value;
            OnExportPathChanged(value);
            OnExportPathChanged(default, value);
            OnPropertyChanged(global::CommunityToolkit.Mvvm.ComponentModel.__Internals.__KnownINotifyPropertyChangedArgs.ExportPath);
            ExportButtonCommand.NotifyCanExecuteChanged();
            Broadcast(__oldValue, value, "ExportPath");
        }
    }
}
```

`Broadcast` is called after `ExportButtonCommand.NotifyCanExecuteChanged();`, therefore the button is disabled and keeps disabled, until the next output change.

Of course this code is simplified and there are other properties that are checked for `CanExport`.

My question is:
  a) Is this an ok-ish way to do?
  b) What are my possibilities to raise a `ExportButtonCommand.NotifyCanExecuteChanged();` from the `ExportService`, so that the button can be enabled again?

**Comment by kg-github** [comment]
Just ran into this exact same issue. I have a CanExecute method for a save button looking at an IsDirty flag on a detail view model. That view model gets updated via a message when an item is selected from a master ListView. Problem is CanExecute is called before the broadcast and so is looking at the previous detail viewmodel state. Even the docs here indicate that Broadcast() should get called before the PropertyChanged() calls:
https://devblogs.microsoft.com/dotnet/announcing-the-dotnet-community-toolkit-800/#broadcast-change-support-for-generated-properties
